herd/valconstraint.ml

```diff
@@ -219,18 +240,81 @@ and type state = A.state =
         let v = subst_atom m v
         and w = subst_atom m w in
         if V.compare v w = 0 then k else Assign (v,Atom w)::k
+    | Assign (v,Deref w) ->
+        let v = subst_atom m v
+        and w = subst_atom m w in
+        if V.compare v w = 0 then k else Assign (v,Deref w)::k
+    | Assign (v,Address w) ->
+        let v = subst_atom m v
+        and w = subst_atom m w in
+        if V.compare v w = 0 then k else Assign (v,Address w)::k
     | Assign (v,e) ->
         let v = subst_atom m v
         and e = subst_expr m e in
         Assign (v,e)::k
-    | Unroll _ -> cn::k
+    | AssignAddr (v,e) ->
+        let v = subst_atom m v
+        and e = subst_expr m e in
+        AssignAddr (v,e)::k
+    | Unroll _ | Inv _ -> cn::k
 
     let subst_cns soln cns = List.fold_right (subst_cn soln) cns []
 
+    let match_addr v1 = function
+      | Inv (AssignAddr (v2, Atom v)) when v1 = v2-> Some v
+      | _ -> None
+
+    let match_loc v1 = function
+      | Inv (AssignAddr(v, Atom v2)) when v1 = v2-> Some v
+      | _ -> None
+
+    let simplify_addr_in_cnstrnt cs c =
+
+      let find_addr a =
+        (* Subst a symbol for a numeric address *)
+        let vs = List.filter_map (match_addr a) cs in
+        if List.length vs == 0 then begin
+          (* If we can't find the address for a, and it is concrete,
+           * or symbolic (but not known) *)
+          if V.is_concrete a || V.is_symbol a then
+            (false, Atom a)
+          else
+            (false, Address a)
+        end else
+          (true, Atom (List.hd vs)) in
+
+      let find_loc a =
+        let vs = List.filter_map (match_loc a) cs in
+        if List.length vs == 0 then
+          if V.is_symbol a then (false, Atom a) else (false, Deref a)
+        else
+          (true, Atom (List.hd vs)) in
+      (* Addresses can have 2 cases *)
+      (* 1. Address of Symbolic value x, for which you want the address 0xf00 *)
+      (* 2. Dereferencing a numeric address 0xf00, we want to get x *)
+      match c with
+        | Assign (v, Address a) ->
+          let (b,a) = find_addr a in (b, Assign (v, a))
+        | Assign (v, Deref p) ->
+          let (b,p) = find_loc p in (b, Assign (v, p))
+        | _ -> (false, c)
+
+    (* This step simplifies variables to their numeric addresses *)
+    (* and dereferences addresses to their symbolic variables *)
+    (* We recursively resolve addresses as much as we can until no changes*)
+    (* are made to the constraints *)
+    let rec simplify_addr_in_cnstrnts cs =
+      let bcs = List.map (simplify_addr_in_cnstrnt cs) cs in
+      if List.exists fst bcs then
+        simplify_addr_in_cnstrnts (List.map snd bcs)
+      else
+        List.map snd bcs
```

Comment:
As I understand the Inv constraint are rigid (_i.e._ do not change). Hence it can be slightly more efficient to extract them from the constraint list before simplifying addresses.
```suggestion
    let simplify_addr_in_cnstrnts cs =
      let invs,cs = List.partition (fun c -> match c with Inv _ -> true | _ -> false) cs in
      
      let rec do_simpiify cs = 
        let bcs = List.map (simplify_addr_in_cnstrnt inv) cs in
        if List.exists fst bcs then
          do_simplify (List.map snd bcs)
      else
        List.map snd bcs in
        
      inv@do_simplify cs
```
Another solution, which would imply more changes, would be to replace the constraint list by a pair or record that would keep invariants and standard constraints separated.

Reply:
good point, I will fix this, but I wonder if the if statement could be expressed in a more idiomatic fashion as well?
I've added your fix and will think about the if